File: src/core/esaignore.ts
import * as fs from 'fs';
import * as path from 'path';

// .esaignore holds one path pattern per line, relative to the file's own directory (the
// project root, by convention — same as .gitignore). Blank lines and lines starting with
// '#' are skipped.
//
// decision: supports exactly two pattern shapes — a literal path/directory (matches
// itself, and everything under it if it's a directory; a bare name with no '/' also
// matches at any depth, like gitignore) and a single-segment basename glob containing '*'
// (e.g. `*.generated.ts`) — not a full gitignore engine (no negation, no '**', no brace
// expansion), matching scan.ts's existing minimal-glob philosophy (see its
// expandGlobLike) rather than pulling in a dependency.
export const ESAIGNORE_FILENAME = '.esaignore';

export function loadIgnorePatterns(rootDir: string): string[] {
    const ignorePath = path.join(rootDir, ESAIGNORE_FILENAME);
    if (!fs.existsSync(ignorePath)) {
        return [];
    }
    return fs
        .readFileSync(ignorePath, 'utf8')
        .split('\n')
        .map((line) => line.trim())
        .filter((line) => line.length > 0 && !line.startsWith('#'))
        .map((line) => line.replace(/\/+$/, ''));
}

function matchesLiteralPattern(relPath: string, pattern: string): boolean {
    if (pattern.includes('/')) {
        return relPath === pattern || relPath.startsWith(`${pattern}/`);
    }
    return relPath.split('/').includes(pattern);
}

function basenameGlobToRegExp(pattern: string): RegExp {
    const escaped = pattern
        .split('*')
        .map((part) => part.replace(/[.+?^${}()|[\]\\]/g, '\\$&'))
        .join('.*');
    return new RegExp(`^${escaped}$`);
}

// Whether `absolutePath` (a file or directory) matches any pattern loaded via
// loadIgnorePatterns, rooted at `rootDir`.
export function isIgnored(absolutePath: string, rootDir: string, patterns: string[]): boolean {
    if (patterns.length === 0) {
        return false;
    }
    const relPath = path.relative(rootDir, absolutePath).split(path.sep).join('/');
    const basename = path.basename(absolutePath);

    return patterns.some((pattern) =>
        pattern.includes('*') ? basenameGlobToRegExp(pattern).test(basename) : matchesLiteralPattern(relPath, pattern)
    );
}

File: src/Core/Esaignore.fs
module Energy.Core.Esaignore

open Fable.Core
open Energy.Core.FsPath
open Energy.Core.Paths

// A `.esaignore` pattern: a literal path or a single-segment basename glob.
//
// decision: typed rather than left as a raw string so a pattern can no longer be transposed with
// the path it is matched against.
// invariant: every `IgnorePattern` value has exactly its wrapped string as its JavaScript
// representation.
[<Erase>]
type IgnorePattern = IgnorePattern of string

let esaignoreFileName = ".esaignore"

// decision: deliberately supports only literal paths and single-segment basename globs; matching
// scan's intentionally small glob surface avoids silently claiming full gitignore semantics.
let loadIgnorePatterns (rootDir: string) : string list =
    let ignorePath = joinPath (Path rootDir) (Path esaignoreFileName)

    if not (existsSync ignorePath) then
        []
    else
        (readFileSync ignorePath (Encoding "utf8")).Split('\n')
        |> Array.map _.Trim()
        |> Array.filter (fun line -> line <> "" && not (line.StartsWith "#"))
        |> Array.map (fun line -> line.TrimEnd('/'))
        |> Array.toList

let private matchesLiteralPattern (relative: string) (IgnorePattern pattern) =
    if pattern.Contains "/" then
        relative = pattern || relative.StartsWith(pattern + "/")
    else
        relative.Split('/') |> Array.contains pattern

let private matchesBasenameGlob (IgnorePattern pattern) (name: string) =
    let pieces = pattern.Split('*')

    let rec loop (remaining: string list) (position: int) =
        match remaining with
        | [] -> position = name.Length
        | first :: rest when position = 0 ->
            if name.StartsWith first then
                loop rest first.Length
            else
                false
        | part :: rest ->
            let index = name.IndexOf(part, position, System.StringComparison.Ordinal)
            if index < 0 then false else loop rest (index + part.Length)

    loop (pieces |> Array.toList) 0

// decision: keeps its Path values unwrapped — they flow straight into the fs/path bindings
// (relativePath/basename) rather than round-tripping through a destructured string.
let isIgnored (absolutePath: Path) (rootDir: Path) (patterns: string list) =
    if patterns.IsEmpty then
        false
    else
        let relative =
            relativePath rootDir absolutePath
            |> fun path -> path.Replace(pathSeparator, "/")

        let name = basename absolutePath

        patterns
        |> List.exists (fun pattern ->
            let typedPattern = IgnorePattern pattern

            if pattern.Contains "*" then
                matchesBasenameGlob typedPattern name
            else
                matchesLiteralPattern relative typedPattern)
